### PR TITLE
Add wal e support

### LINKED ISSUE #119: Add Wal-e backup support for PostgreSQL
add lines : 

```
archive_mode = on
archive_command = 'envdir /var/lib/postgresql/data/wal-e.d/env wal-e wal-push %p'
archive_timeout = 60
```

in REPLICATION

### PR DESCRIPTION
**Pull request by joshuaberetta**
Closes #119 

The main changes have been made to `config.py` and `postgres.conf.tpl` with minor changes to the docker-compose templates for both the primary and secondary override files.

Within `config.py`, the default value for `postgres_settings_content` was set to ensure that settings are always carried through to `kobo-env/postgres/primary/postgres.conf` and `kobo-env/postgres/secondary/postgres.conf`. I noticed that when there is a previous value set in `.run.conf`, the `postgres_settings_content` is not updated, so I forced this to update to the default value if the user answers "no" to the question, "Do you want to tweak PostgreSQL settings?".

If the user answers "yes" to "Do you want to activate backups?" and then "yes" to "Do you want to use WAL-E for continuous archiving of PostgreSQL backups?", they will not be asked about setting `backup_from_primary` or the cron schedule set to `postgres_backup_schedule`, `backup_from_primary` is then set to TRUE by default. If the user answers "no" to WAL-E backups, the environment variable `USE_WAL_E_BACKUPS` will be set to "#" and will comment out the code for WAL-E in the `kobo-env/postgres/primary/postgres.conf` file, and the environment variable `USE_WAL_E` is set to FALSE.

In the two docker-compose template files for primary and secondary, the environment variable `OVERRIDE_POSTGRES_SETTINGS` is no longer considered, and the volume is mounted by default.

### CODE REVIEW

File: helpers/template.py [hunk 1]
```diff
@@ -215,6 +215,8 @@ def _get_value(property_, true_value="", false_value="#",
             "REDIS_MAIN_PORT": config.get("redis_main_port", "6739"),
             "REDIS_CACHE_PORT": config.get("redis_cache_port", "6380"),
             "USE_BACKUP": "" if config.get("use_backup") == Config.TRUE else "#",
+            "USE_WAL_E_BACKUP": "" if config.get("use_wal_e") == Config.TRUE else "#",
```
Comment (noliveleger) [inline helpers/template.py:218]: Would you mind to use `_get_value` instead? 
I've noticed that several lines should be updated too ;-( We should open a separate issue for that.

File: templates/kobo-env/postgres/primary/postgres.conf.tpl [hunk 2]
```diff
@@ -14,3 +14,7 @@
 # Total Memory (RAM): ${POSTGRES_RAM}GB
 
 ${POSTGRES_SETTINGS}
+
+${USE_WAL_E_BACKUP}archive_mode = on
+${USE_WAL_E_BACKUP}archive_command = 'envdir /var/lib/postgresql/data/wal-e.d/env wal-e wal-push %p'
```
Comment (noliveleger) [inline templates/kobo-env/postgres/primary/postgres.conf.tpl:19]: I think `/var/lib/postgresql/data` could be replaced by `$PGDATA`.

**Review by noliveleger — CHANGES_REQUESTED** [review]

File: helpers/template.py (hunk 1 above)
Comment (joshuaberetta) [reply helpers/template.py:218]: No problem! Should I update all instances of `config.get()` with `_get_value()`?

File: helpers/template.py [hunk 3]
```diff
@@ -215,7 +215,9 @@ def _get_value(property_, true_value="", false_value="#",
             "REDIS_MAIN_PORT": config.get("redis_main_port", "6739"),
             "REDIS_CACHE_PORT": config.get("redis_cache_port", "6380"),
             "USE_BACKUP": "" if config.get("use_backup") == Config.TRUE else "#",
-            "USE_WAL_E_BACKUP": "" if config.get("use_wal_e") == Config.TRUE else "#",
+            "USE_WAL_E_BACKUP": _get_value("use_wal_e", true_value="",
```
Comment (noliveleger) [inline helpers/template.py:218]: default values for `_get_value()` are:
- `true_value=""`
- `false_value="#"`,
- `comparison_value=Config.TRUE`

For this particular case, you could just call `_get_value('use_wal_e')`

**Review by noliveleger — CHANGES_REQUESTED** [review]

File: templates/kobo-env/postgres/primary/postgres.conf.tpl (hunk 2 above)
Comment (rroux) [reply templates/kobo-env/postgres/primary/postgres.conf.tpl]: Can't use variable directly in postgresql.con file
example:
`archive_command = 'envdir $PGDATA/wal-e.d/env wal-e wal-push% p'`

It works with a sed when starting the docker:
`sed -i '/archive_mode/ a\'"archive_command = 'envdir $PGDATA/wal-e.d/env wal-e wal-push %p'" /var/lib/postgresql/data/postgresql.conf`

Comment (joshuaberetta) [reply templates/kobo-env/postgres/primary/postgres.conf.tpl]: I got it to work with @noliveleger's suggestion of using an escape character in the `postgres.conf.tpl` file: `${USE_WAL_E_BACKUP}archive_command = 'envdir $$PGDATA/wal-e.d/env wal-e wal-push %p'`

Or are you saying that `$PGDATA` can't actually be used directly in the resulting `postgresql.conf` file?

**Review by noliveleger — COMMENTED** [review]
Would mind to bump version to `3.4.0` and use branch `add-wale-support` over [here]
(https://github.com/kobotoolbox/kobo-install/pull/122/files#diff-3a81b92b1ad3cbc4cda6c831dd395919345ec5398ae1e97eae2e0dd894552d04R35-R36)

File: templates/kobo-env/postgres/primary/postgres.conf.tpl (hunk 2 above)
Comment (noliveleger) [reply templates/kobo-env/postgres/primary/postgres.conf.tpl]: The substitution of `PGDATA` is done in `postgres` container. See https://github.com/kobotoolbox/kobo-docker/pull/310/

**Review by noliveleger — CHANGES_REQUESTED** [review]

File: readme.md [hunk 4]
```diff
@@ -110,6 +110,7 @@ User can choose between 2 types of installations:
 |MongoDB user's password|  **Autogenerate**  | ✓ | ✓ |
 |Redis password<sup>4</sup>|  **Autogenerate**  | ✓ | ✓ |
 |Use AWS storage|  **No**  | ✓ | ✓ (frontend only) |
+|Use WAL-E PostgreSQL backups<sup>5</sup> |  **No**  | ✓ | ✓ (frontend only) |
```
Comment (noliveleger) [inline readme.md:113]: I think it should be `backend only`. 
BTW, would you mind to remove (frontend only) for AWS storage. We do ask for AWS credentials in the backend if backups are activated (maybe another exponent <sup>n</sup> could explain it)

**Review by noliveleger — CHANGES_REQUESTED** [review]

File: helpers/config.py [hunk 5]
```diff
@@ -774,10 +810,10 @@ def __questions_backup(self):
                                 "0 0 * * 0"))
 
                     if self.backend_questions:
-
+                        
                         if self.primary_backend:
                             CLI.colored_print("Run PostgreSQL backup from primary backend server?",
-                                              CLI.COLOR_SUCCESS)
+                                            CLI.COLOR_SUCCESS)
```
[the remark this answers is not in the document]
Comment (noliveleger) [reply helpers/config.py:816]: There is also an issue when running from the secondary back-end server.
User is not prompted for the backup schedule 

```
AWS Access Key: xxxxx
AWS Secret Key: xxxx
AWS Bucket name: xxxxx
╔═════════════════════════════════════════════════════════════════╗
║ Schedules use linux cron syntax with UTC datetimes.             ║
║ For example, schedule at 12:00 AM E.S.T every Sunday would be:  ║
║ 0 5 * * 0                                                       ║
║                                                                 ║
║ Please visit https://crontab.guru/ to generate a cron schedule. ║
╚═════════════════════════════════════════════════════════════════╝
AWS Backups bucket name [xxxxx]: xxxxx
How many yearly backups to keep?
```

If should be

```
AWS Access Key: xxxxx
AWS Secret Key: xxxx
AWS Bucket name: xxxxx
╔═════════════════════════════════════════════════════════════════╗
║ Schedules use linux cron syntax with UTC datetimes.             ║
║ For example, schedule at 12:00 AM E.S.T every Sunday would be:  ║
║ 0 5 * * 0                                                       ║
║                                                                 ║
║ Please visit https://crontab.guru/ to generate a cron schedule. ║
╚═════════════════════════════════════════════════════════════════╝
PostgreSQL backup schedule?
[0 2 * * 0]:
AWS Backups bucket name [xxxxx]: xxxxx
How many yearly backups to keep?
```

**Review by noliveleger — CHANGES_REQUESTED** [review]

File: helpers/config.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (joshuaberetta) [reply helpers/config.py:816]: I'm not actually sure why that indent was removed on my side... but I'll keep that in mind.

[the remark this answers is not in the document]
Comment (joshuaberetta) [reply helpers/config.py:816]: > This block is executed even if I chose Use WAL-E backups. It should not, WAL-E cannot be run from the secondary.

I thought this was what you mentioned in an earlier comment here, so I changed to to make it asked for the schedule.

> We need this info to setup the cron task within PostgreSQL container.

What were you referring to that needed to be changed?

[the remark this answers is not in the document]
Comment (joshuaberetta) [reply helpers/config.py:816]: > There is also an issue when running from the secondary back-end server.

I'll get on that 👍

[the remark this answers is not in the document]
Comment (noliveleger) [reply helpers/config.py:816]: > I thought this was what you mentioned in an earlier comment here, so I changed it to make it asked for the schedule.

GH did not make me crystal clear. Sorry about that :-( 

If user chooses WAL-E backups, it implies to run the backup from the primary back-end server. 
So the question 
```
Run PostgreSQL backup from primary backend server?
	1) Yes
	2) No
```
should not asked but 
```
PostgreSQL backup schedule?
[0 2 * * 0]:
```
should.

[the remark this answers is not in the document]
Comment (joshuaberetta) [reply helpers/config.py:816]: I think I found the issue behind your observation:

> There is also an issue when running from the secondary back-end server.
> User is not prompted for the backup schedule

This is how things currently stand:

```python
backup_from_primary = self.__config["backup_from_primary"] == Config.TRUE
if (not self.multi_servers or
      (self.primary_backend and backup_from_primary) or
          (self.secondary_backend and not backup_from_primary)):
               CLI.colored_print("PostgreSQL backup schedule?", CLI.COLOR_SUCCESS)
                            self.__config["postgres_backup_schedule"] = CLI.get_response(
                                "~{}".format(schedule_regex_pattern),
                                self.__config.get(
                                    "postgres_backup_schedule",
                                    "0 2 * * 0"))
```

but the default config dict has this: `"backup_from_primary": Config.TRUE`, therefore the condition 

```python
(self.secondary_backend and not backup_from_primary)
``` 

will always be False and the schedule is never asked even if `secondary_backend` is True. From what I can see, this has been the case for a while.

Does this seem reasonable or am I missing something?

[the remark this answers is not in the document]
Comment (noliveleger) [reply helpers/config.py:816]: @joshuaberetta, You are right! One condition is missing when using the secondary backend.
Maybe [this block](https://github.com/kobotoolbox/kobo-install/blob/34c434d6bf9c95f9c5d59bde2a1f7dc4035d169d/helpers/config.py#L814-L821) above could be something like 
```
                if self.__config['use_wal_e'] == Config.TRUE:
                    self.__config["backup_from_primary"] = Config.FALSE
                else:
                        if self.primary_backend:
                            CLI.colored_print("Run PostgreSQL backup from primary backend server?",
                                            CLI.COLOR_SUCCESS)
                            CLI.colored_print("\t1) Yes")
                            CLI.colored_print("\t2) No")
                            self.__config["backup_from_primary"] = CLI.get_response(
                                [Config.TRUE, Config.FALSE],
                                self.__config.get("backup_from_primary", Config.TRUE))
                      else:
                            self.__config["backup_from_primary"] = Config.FALSE
```

Be aware, I have **not** tested this snippet

File: helpers/config.py [hunk 6]
```diff
@@ -811,14 +812,20 @@ def __questions_backup(self):
...
-                        if self.primary_backend:
-                            CLI.colored_print("Run PostgreSQL backup from primary backend server?",
-                                            CLI.COLOR_SUCCESS)
-                            CLI.colored_print("\t1) Yes")
-                            CLI.colored_print("\t2) No")
-                            self.__config["backup_from_primary"] = CLI.get_response(
-                                [Config.TRUE, Config.FALSE],
-                                self.__config.get("backup_from_primary", Config.TRUE))
+                        if self.__config['use_wal_e'] == Config.TRUE:
+                            self.__config["backup_from_primary"] = Config.FALSE
```
Comment (noliveleger) [inline helpers/config.py:816]: You should not have copied my mistakes :P 😇 

Backup is enforce on primary with WAL-E 

```
self.__config["backup_from_primary"] = Config.TRUE 
```

File: helpers/config.py [hunk 7]
```diff
@@ -811,14 +812,20 @@ def __questions_backup(self):
...
+                            if self.primary_backend:
+                                CLI.colored_print("Run PostgreSQL backup from primary backend server?",
+                                                CLI.COLOR_SUCCESS)
+                                CLI.colored_print("\t1) Yes")
+                                CLI.colored_print("\t2) No")
+                                self.__config["backup_from_primary"] = CLI.get_response(
+                                    [Config.TRUE, Config.FALSE],
+                                    self.__config.get("backup_from_primary", Config.TRUE))
+                            else:
+                                print('I got here')
```
Comment (noliveleger) [inline helpers/config.py:827]: 👀

**Review by noliveleger — CHANGES_REQUESTED** [review]

**Comment by noliveleger** [comment]
LGTM.

**Review by noliveleger — APPROVED** [review]

File: helpers/config.py (hunk 7 above)
Comment (joshuaberetta) [reply helpers/config.py]: Haha oh dear!

**Pull request merged**